transport/oob/oob.go

```diff
@@ -0,0 +1,137 @@
+	w.oobPosition -= written
+	return written, err
+}
+
+func (w *oobWriter) Write(data []byte) (int, error) {
+	var written int
+	var err error
+
+	if conn, ok := w.writer.(net.Conn); ok {
+		if w.oobPosition > 0 && w.oobPosition < int64(len(data)) {
+			// Write the first part with the regular writer
+			written, err = w.writer.Write(data[:w.oobPosition])
+			if err != nil {
+				return written, err
+			}
+
+			// Send the specified OOB byte using the new sendOOBByte method
+			_ = conn
+			err = w.sendOOBByte(conn)
+			if err != nil {
+				return written, err
+			}
+			data = data[written:] // Skip the OOB byte
+		}
+	}
+	// Write the remaining data
+	n, err := w.writer.Write(data)
+	written += n
+	return written, err
+}
+
+// sendOOBByte sends the specified OOB byte over the provided connection.
+// It sets the appropriate flags based on whether disOOB mode is enabled.
+func (w *oobWriter) sendOOBByte(conn net.Conn) error {
+	// Attempt to convert to *net.TCPConn to access SyscallConn
+	tcpConn, ok := conn.(*net.TCPConn)
+	if !ok {
+		return errors.New("connection is not a TCP connection")
+	}
+
+	oobData := []byte{w.oobByte}
+	var flags int
+	if w.disOOB {
+		flags = syscall.MSG_OOB | syscall.MSG_DONTROUTE // Additional flag for disOOB mode
+	} else {
+		flags = syscall.MSG_OOB
+	}
+
+	// Use SyscallConn to access the underlying file descriptor safely
+	rawConn, err := tcpConn.SyscallConn()
+	if err != nil {
+		return err
+	}
+
+	// Use Control to execute Sendto on the file descriptor
+	var sendErr error
+	err = rawConn.Control(func(fd uintptr) {
+		sendErr = syscall.Sendto(int(fd), oobData, flags, nil)
+		if sendErr != nil {
+			fmt.Errorf("failed to send OOB byte: %v", sendErr)
+		}
+	})
+	if err != nil {
+		return err
+	}
+	return sendErr
+}
+
+// oobDialer is a dialer that applies the OOB and disOOB strategies.
+type oobDialer struct {
+	dialer      transport.StreamDialer
+	oobPosition int64
+	oobByte     byte
+	disOOB      bool
+}
+
+// NewStreamDialerWithOOB creates a [transport.StreamDialer] that applies OOB byte sending at "oobPosition" and supports disOOB.
+// "oobByte" specifies the value of the byte to send out-of-band.
+func NewStreamDialerWithOOB(dialer transport.StreamDialer, oobPosition int64, oobByte byte, disOOB bool) (transport.StreamDialer, error) {
+	if dialer == nil {
+		return nil, errors.New("argument dialer must not be nil")
+	}
+	return &oobDialer{dialer: dialer, oobPosition: oobPosition, oobByte: oobByte, disOOB: disOOB}, nil
+}
+
+// DialStream implements [transport.StreamDialer].DialStream with OOB and disOOB support.
+func (d *oobDialer) DialStream(ctx context.Context, remoteAddr string) (transport.StreamConn, error) {
+	innerConn, err := d.dialer.DialStream(ctx, remoteAddr)
+	if err != nil {
+		return nil, err
+	}
+	// Wrap connection with OOB and/or disOOB writer based on configuration
+	return transport.WrapConn(innerConn, innerConn, NewOOBWriter(innerConn, d.oobPosition, d.oobByte, d.disOOB)), nil
```

Comment:
You should test if `innerConn` is a `*net.TCPConn` here, before wrapping.

Reply:
added